pkg/reconciler/ingress/ingress.go

```diff
@@ -0,0 +1,133 @@
+	if err != nil {
+		logger.Errorf("error reconciling ingress: %s", desired.Name)
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileIngress(ctx context.Context, desired *v1alpha1.Ingress) (*v1alpha1.Ingress, error) {
+	desired.Status.InitializeConditions()
+	ingress, err := r.ingressLister.Ingresses(desired.Namespace).Get(desired.Name)
+	if apierrs.IsNotFound(err) {
+		ingress, err = r.netclient.NetworkingV1alpha1().Ingresses(desired.Namespace).Create(ctx, desired, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create Ingress: %w", err)
+		}
+		return ingress, nil
+	} else if err != nil {
+		return nil, err
+	} else if !equality.Semantic.DeepEqual(ingress.Spec, desired.Spec) ||
+		!equality.Semantic.DeepEqual(ingress.Annotations, desired.Annotations) {
+		// Don't modify the informers copy
+		origin := ingress.DeepCopy()
+		origin.Spec = desired.Spec
+		origin.Annotations = desired.Annotations
+		updated, err := r.netclient.NetworkingV1alpha1().Ingresses(origin.Namespace).Update(ctx, origin, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to update Ingress: %w", err)
+		}
+		return updated, nil
+	}
+	return ingress, err
+}
+
+// makeNewIngress creates an Ingress object with respond-async headers pointing to producer-service
+func makeNewIngress(ingress *v1alpha1.Ingress, ingressClass string) *v1alpha1.Ingress {
+	original := ingress.DeepCopy()
+	splits := make([]v1alpha1.IngressBackendSplit, 0, 1)
+	splits = append(splits, v1alpha1.IngressBackendSplit{
+		IngressBackend: v1alpha1.IngressBackend{
+			ServiceName:      "producer-service", // TODO(beemarie): make this configurable
+			ServiceNamespace: "default",
+			ServicePort:      intstr.FromInt(80),
+		},
+		Percent: int(100),
+	})
+	theRules := []v1alpha1.IngressRule{}
+	for _, rule := range original.Spec.Rules {
+		newRule := rule
+		newPaths := make([]v1alpha1.HTTPIngressPath, 0)
+		newPaths = append(newPaths, v1alpha1.HTTPIngressPath{
+			Headers: map[string]v1alpha1.HeaderMatch{"Prefer": {Exact: "respond-async"}},
+			Splits:  splits,
+		})
+		newPaths = append(newPaths, newRule.HTTP.Paths...)
+		newRule.HTTP.Paths = newPaths
+		theRules = append(theRules, newRule)
+	}
+	return &v1alpha1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      original.Name + "-new",
+			Namespace: original.Namespace,
+			Annotations: kmeta.FilterMap(kmeta.UnionMaps(map[string]string{
+				networking.IngressClassAnnotationKey: ingressClass,
+			}), func(key string) bool {
+				return key == corev1.LastAppliedConfigAnnotation
+			}),
+			Labels:          original.Labels,
+			OwnerReferences: original.OwnerReferences,
+		},
+		Spec: v1alpha1.IngressSpec{
+			Rules: theRules,
+		},
+	}
+}
+
+// TODO(beemarie) track status of upstream ingress that is created "-new"
+func markIngressReady(ingress *v1alpha1.Ingress) {
+	internalDomain := domainForServiceName(ingress.Name)
+	externalDomain := domainForServiceName(ingress.Name)
```

Comment:
nit: if these are the same you might want to avoid calling the function twice

Reply:
meh, ignore me, this will disappear anyway when we propagate properly won't it?

Reply:
Yes - this will be a part of the propagation work.